rfc/rfc-0014-binding-injection.md

```diff
@@ -0,0 +1,131 @@
+# RFC-0014: Binding Injection
+
+**Authors:** @scothis
+
+**Status:**
+
+**Pull Request URL:** [#1382](https://github.com/projectriff/riff/pull/1382)
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** [RFC 0002 Bindings](./rfc-0002-bindings.md), [RFC 0008 Pod Templates](./rfc-0008-pod-templates.md)
+
+
+## Problem
+
+[RFC 0002 Bindings](./rfc-0002-bindings.md) defines a interface for a resource to expose metadata and secrets to be consumed by another resource. It leverages the [CNB Binding](https://github.com/buildpack/spec/blob/master/extensions/bindings.md) spec for how the service metadata and secrets are exposed inside a container. While a binding can be consumed manually, it's difficult to manage as it requires defining two volumes, and two volume mounts and a environment variable per container.
+
+### Anti-Goals
+
+This RFC has no opinion as to the type of binding being injected, or the consumer of the binding.
+
+## Solution
+
+Knative Bindings (yea, the term is overloaded) define an abstract way for reconcilers to operate on resources by duck typing known elements of the resource without requiring pre-knowledge or full type knowledge at runtime. By leveraging resources that use a PodTemplateSpec, it's possible to inject volumes, volume mounts and environment variables into an existing resource (even if owned and controlled by another resource). With [RFC 0008 Pod Templates](./rfc-0008-pod-templates.md) the riff Deployers and Processor resources implement the PodSpecable duck type.
+
+A duck type for exposing a binding is defined in RFC 0002. While not as wide spread, the Stream resource currently implements this duck type. Combining the Binding and PodSpecable duck types allows for a generic means to inject binding metadata and secrets into a consuming workload. A new CRD can express this injection:
+
+```yaml
+apiVersion: bindings.projectriff.io/v1alpha1
+kind: Injection
+metadata:
+  name: my-binding-injection
+spec:
+  binding:
+    apiVersion: streaming.projectriff.io/v1alpha1
+    kind: Stream
+    name: my-stream
+  containers:
+  - user-container
+  subject:
+    apiVersion: knative.projectriff.io/v1alpha1
+    kind: Deployer
+    name: my-deployer
+status:
+  observedGeneration: 1
+  conditions: []
+```
+
+- `.spec.binding` is an ObjectReference to a resource implementing the binding duck type
+- `.spec.containers` optionally restricts which containers are augmented, defaults to all containers
+- `.spec.subject` is an ObjectReference to a resource implementing the podspecable duck type
```

Comment:
I know it's already too late / out of scope to reconsider the word "binding" but to me, the "binding" is the act to bind, so would rather map to what you call "injection".

I also have issues with the word "subject" though. Subject to me means something that is "acting". Here, the subject is more akin to a "destination", something _where_ the injection happens.

The combination of ambiguity on both "binding" and "subject" would make this mechanism hard to use IMO, never knowing which is which. The fact that you use the word "target" below is interesting too, as I would reserve that word for what is here called the subject, because this conveys the idea of injecting _into_.

Reply:
Ditto (I had the same thought about "target").

Reply:
> I also have issues with the word "subject" though. Subject to me means something that is "acting". Here, the subject is more akin to a "destination", something where the injection happens.

`subject` comes from Knative Bindings, which we should align with. I'm still not fully up to speed on the exact semantics, so the the shape of the resource will likely evolve as we better understand the implementation. This is the main reason why this RFC is a draft.